esmvaltool/diag_scripts/shared/_base.py

```diff
@@ -51,14 +53,85 @@ def get_diagnostic_filename(basename, cfg, extension='nc'):
     -------
     str:
         A valid path for saving a diagnostic data file.
-
     """
     return os.path.join(
         cfg['work_dir'],
         f"{basename}.{extension}",
     )
 
 
+def save_figure(basename, provenance, cfg, figure=None, close=True, **kwargs):
+    """Save a figure to file.
+
+    Parameters
+    ----------
+    basename: str
+        The basename of the file.
+    provenance: dict
+        The provenance record for the figure.
+    cfg: dict
+        Dictionary with diagnostic configuration.
+    figure: matplotlib.figure.Figure
+        Figure to save.
+    close:
+        Close the figure after saving.
+    **kwargs:
+        Keyword arguments to pass to :obj:`matplotlib.figure.Figure.savefig`.
+
+    See Also
+    --------
+    ProvenanceLogger: For an example provenance record that can be used
+        with this function.
+    """
+    if cfg.get('output_file_type') is None:
+        extensions = ('png', 'pdf')
+    elif isinstance(cfg['output_file_type'], str):
+        extensions = (cfg['output_file_type'], )
+    else:
+        extensions = cfg['output_file_type']
+
+    for ext in extensions:
+        filename = Path(cfg['plot_dir']) / ext / f"{basename}.{ext}"
+        filename.parent.mkdir(exist_ok=True)
+        logger.info("Plotting analysis results to %s", filename)
+        fig = plt if figure is None else figure
+        fig.savefig(filename, **kwargs)
+        with ProvenanceLogger(cfg) as provenance_logger:
+            provenance_logger.log(filename, provenance)
+
+    if close and figure:
+        plt.close(figure)
+
+
+def save_data(basename, provenance, cfg, cube, **kwargs):
+    """Save the data used to create a plot to file.
+
+    Parameters
+    ----------
+    basename: str
+        The basename of the file.
+    provenance: dict
+        The provenance record for the data.
+    cfg: dict
+        Dictionary with diagnostic configuration.
+    cube: iris.cube.Cube
+        Data cube to save.
+    **kwargs:
+        Keyword arguments to pass to :obj:`iris.save`.
+
+    See Also
+    --------
+    ProvenanceLogger: For an example provenance record that can be used
+        with this function.
+    """
+    if cfg.get('write_netcdf', True):
```

Comment:
This will be removed when `write_netcdf` gets deprecated, right? In this case I think you should also add `write_plots` to `save_figure`.

Reply:
No, I did not intent to remove this later. My idea here was that it would be nice to still be able to specify `write_netcdf` in the recipe (rather than in the config-user file), because if the output netcdf files are very large people might like this. However, for most recipes this probably doesn't make sense, since typically the plotted data is small and there is no performance gain whatsoever in disabling writing netcdf files.

Reply:
So would it make sense to add a similar line to `save_figure` with `write_plots`?

Reply:
I'm not really convinced that option is useful for anything, can you imagine a situation where writing plots needs to be disabled because there is considerable gain in computation time or disk space use? You mentioned writing PDF files with many points in them, but that can easily be solved by choosing a different output file type already.

Reply:
Me neither, but why would be support disabling `write_netcdf`? I think we should either support both options or none.

Reply:
Because for `write_netcdf` I can see that if a diagnostic produces very large output files, some people may not want this. For example, @ledm has not implemented writing output netcdf files at all for some of his diagnostics because he feels it's just copying the large preprocessor output. However, I cannot think of a single case where not writing the plots offers any advantage.